### PR TITLE
Cli

### PR DESCRIPTION
**Pull request by amiller68**
Starting a PR to merge my work on Implementing a general Purpose EthClient for all our rust eth needs + Prep work defining what's a deal Proposal in rust. 

EthClient comes with a builder that can be used to
- Initialize an Eth Client for ENV vars
- or using a custom configuration
EthClient can be used to hold:
  - a provider (required)
  - a chain_id (required)
  - a signer (optional)
  - a contract (optional)
  - a timeout (optional)
Which defines everything needed to talk to Eth through a specified API (Infura by default)

### CODE REVIEW

**Comment by amiller68** [comment]
Fixed Merge Conflicts and got my tests to pass.

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /target
-
+.idea # IntelliJ IDEA
```
Comment (laudiacay) [inline .gitignore:2]: your whole .idea is still committed, please run this: https://stackoverflow.com/questions/7927230/remove-directory-from-remote-repository-after-adding-them-to-gitignore

also, in the future, make a Draft PR, then go to "files changed" and check that everything's ready to go, and then tag a reviewer please! (helps with notification spam)

File: README.md [hunk 2]
```diff
@@ -1,3 +1,8 @@
 # banyan shared repos
 
 you need to put the codecov badge here
+
+
+# Testing 
+
+- Remember to set ENV variables for testing
```
Comment (laudiacay) [inline README.md:8]: make this a TODO so it gets picked up by the CI hook thing

File: Cargo.toml [hunk 3]
```diff
@@ -10,15 +10,17 @@ edition = "2021"
 # TODO better error types than anyhow
 anyhow = "1.0"
 ethers = { git = "https://github.com/gakonst/ethers-rs" }
+ethers-contract = "0.17.0"
```
Comment (laudiacay) [inline Cargo.toml:13]: maybe we should have a monorepo and stick ipfs-proof-buddy in here too

File: src/eth.rs [hunk 4]
```diff
@@ -1,75 +1,163 @@
...
+
+    /* Struct State Methods */
+
+    /// Return whether theres's a signer configured
+    pub fn has_signer(&self) -> bool {
+        self.signer.is_some()
+    }
+
+    /// Return whether theres's a contract configured
+    pub fn has_contract(&self) -> bool {
```
Comment (laudiacay) [inline src/eth.rs:75]: why aren't these just initialized once in a new()?

File: src/eth.rs [hunk 5]
```diff
@@ -1,75 +1,163 @@
...
+    // #[indexed]
+    pub executor: Address,
+    // #[indexed]
+    pub offerId: U256, // TODO: Snake case
+}
+
+/// The EthProvider is a wrapper around the ethers-rs Provider that handles all Ethereum interactions.
+impl EthClient {
+    /// Create a new default EthClientBuilder
+    pub fn builder() -> EthClientBuilder {
```
Comment (laudiacay) [inline src/eth.rs:63]: why

File: src/eth.rs [hunk 6]
```diff
@@ -1,75 +1,163 @@
...
+        // Borrow our signer and contract
+        let signer = self.signer.as_ref().unwrap();
+        let contract = self.contract.as_ref().unwrap();
+        // TODO: Tokenize the DealProposal
+        dbg!("DP: {:?}", &dp);
+
+        dbg!("Tokenizing Arguments");
+        let data = contract.encode("startOffer", dp).unwrap();
+        // TODO: Create a transaction request
+        dbg!("Initializing mew Request");
```
Comment (laudiacay) [inline src/eth.rs:100]: typo

File: src/eth.rs [hunk 7]
```diff
@@ -1,75 +1,163 @@
...
+        let _tx = TransactionRequest::new()
+            .to(contract.address())
+            .data(data)
+            .gas(10_000_000u64)
```
Comment (laudiacay) [inline src/eth.rs:104]: hmmm should this be hardcoded

File: src/eth.rs [hunk 8]
```diff
@@ -1,75 +1,163 @@
...
+            .chain_id(self.chain_id);
+        // TODO: Sign the transaction and Listen for the event
+        dbg!("Signing Request");
+        let pending_tx = signer.send_transaction(_tx, None).await?;
+        let receipt = pending_tx.await?;
+        dbg!("Receipt: {:?}", receipt.as_ref());
+        let bn = receipt.as_ref().unwrap().block_number.unwrap();
+        let logs: Vec<NewOffer> = contract.event().from_block(bn).query().await?;
+        dbg!("Logs: {:?}", &logs);
+        let event = logs[0].clone();
```
Comment (laudiacay) [inline src/eth.rs:115]: is this right? idk?

File: src/eth.rs [hunk 9]
```diff
@@ -1,75 +1,163 @@
...
+    }
+
+    /* Chain Primitives */
+
+    /// Get the current block number
     pub async fn get_latest_block_num(&self) -> Result<BlockNum> {
-        let provider = self.provider.lock().await;
-        let block = provider.get_block_number().await?;
+        let provider = &self.provider;
+        let block = timeout(self.timeout, provider.get_block_number()).await??;
```
Comment (laudiacay) [inline src/eth.rs:145]: leave timeouts out, i removed them earlier today

File: src/eth.rs [hunk 10]
```diff
@@ -78,7 +166,10 @@ impl VitalikProvider {
         deal_id: DealID,
         window_num: u64,
     ) -> Result<BlockNum> {
-        let contract = self.contract.lock().await;
+        if !self.has_contract() {
+            return Err(anyhow!("No contract available"));
```
Comment (laudiacay) [inline src/eth.rs:170]: this isn't necessary the contract should just be initialized at the start and if it fails it fails

File: src/eth.rs [hunk 11]
```diff
@@ -88,9 +179,53 @@ impl VitalikProvider {
         Ok(res)
     }
 
+    pub async fn u256_method(
```
Comment (laudiacay) [inline src/eth.rs:182]: take this back out please

File: src/eth.rs [hunk 12]
```diff
@@ -173,14 +308,180 @@ impl VitalikProvider {
...
+        // Try and Read the Contract Address from the environment. Default to None if not set
+        let contract_address = match env::var("ETH_CONTRACT_ADDRESS").ok() {
+            Some(address) => Some(address.parse::<Address>().unwrap()),
+            None => None,
+        };
+        dbg!(format!("  CONTRACT_ADDRESS: {:?}", &contract_address));
+
+        // TODO: Add param for ABI
+        Self {
+            api_url,
```
Comment (laudiacay) [inline src/eth.rs:361]: this just feels really really unnecessary

Comment (laudiacay) [reply src/eth.rs:361]: like this whole builder situation, i'm gonna veto sorry

File: src/eth.rs [hunk 13]
```diff
@@ -173,14 +308,180 @@ impl VitalikProvider {
...
+            panic!("No contract available!");
+        }
+        // Try and get the current block number
+        let block_num = eth_client.get_latest_block_num().await.unwrap();
+        println!("Latest Block Number: {}", block_num.0);
+        return;
+    }
+
+    #[tokio::test]
+    /// Test sending a deal Proposal
```
Comment (laudiacay) [inline src/eth.rs:475]: good, but add more tests?

File: src/types.rs [hunk 14]
```diff
@@ -39,6 +49,7 @@ where
     Ok(bao::Hash::from(hash_bytes))
 }
 
+/// DealIds - The onChain ID of a deal submitted to Ethereum
```
Comment (laudiacay) [inline src/types.rs:52]: all these docs are nice :D thank you

File: src/types.rs [hunk 15]
```diff
@@ -90,12 +102,102 @@ impl Mul<usize> for BlockNum {
...
+    pub proof_frequency_in_blocks: u64,
+    /// The amount of tokens to be paid to the executor per TiB, as a float
+    pub price_per_tib: f64,
+    /// The amount of tokens in collateral the executor must provide per TiB, as a float
+    pub collateral_per_tib: f64,
+    /// The Address of the token to use as denominator for the price and collateral, as a string
+    pub erc20_token_denomination: String,
+}
+
+// TODO (al): Don't we need to add executor / creator address to this?
```
Comment (laudiacay) [inline src/types.rs:199]: yes you are correct. this needs to be done in solidity as well. @jonahkaye @xBalbinus

**Review by laudiacay — DISMISSED** [review]

File: Cargo.toml (hunk 3 above)
Comment (amiller68) [reply Cargo.toml:13]: I feel like building core tools, and then implementations on top in separate repos is the way to go. Like maybe ipfs_proof_buddy could define both main and lib, and be importable that way, but why pack a whole Rocket server into a repo that's only gonna get bigger.

Comment (amiller68) [reply Cargo.toml:13]: That being said if there's anything generalizable in what u do have in proof_buddy, lets incorporate them.

**Comment by laudiacay** [comment]
lets move the CLI out then

File: src/eth.rs (hunk 4 above)
Comment (amiller68) [reply src/eth.rs:75]: They're just more readable redeclarations of `is_some` I think, but I could just define them in the struct if u want.

File: src/eth.rs (hunk 5 above)
Comment (amiller68) [reply src/eth.rs:63]: The builder pattern is cool. it let's you define different types of cusomization and initialization logic, and keeps your actual implementations clean. I found it made it easier to reason/type up all this biolerplate.

I can add a `new()` implementation if u want.

File: .gitignore (hunk 1 above)
Comment (amiller68) [reply .gitignore:2]: removed. sorry about that.

File: README.md (hunk 2 above)
Comment (amiller68) [reply README.md:8]: Done

File: src/eth.rs (hunk 5 above)
Comment (amiller68) [reply src/eth.rs:63]: Implemented

File: src/eth.rs (hunk 7 above)
Comment (amiller68) [reply src/eth.rs:104]: prolly not

File: src/eth.rs (hunk 6 above)
Comment (amiller68) [reply src/eth.rs:100]: wanna complete that pokedex tho. (fixed)

File: src/eth.rs (hunk 7 above)
Comment (amiller68) [reply src/eth.rs:104]: I made those configurable on a per call basis.

File: src/eth.rs (hunk 8 above)
Comment (amiller68) [reply src/eth.rs:115]: I am not sure. I have to test when I find a working token.

File: src/eth.rs (hunk 9 above)
Comment (amiller68) [reply src/eth.rs:145]: Ok.

File: src/eth.rs (hunk 10 above)
Comment (amiller68) [reply src/eth.rs:170]: My thinking was you might not need a contract for lots of chain specific view functions, so I thought it aided usability later on.

File: src/types.rs (hunk 14 above)
Comment (amiller68) [reply src/types.rs:52]: ur welcome

File: src/eth.rs (hunk 13 above)
Comment (amiller68) [reply src/eth.rs:475]: willl do -- once I got working with the real contracts working.

File: src/eth.rs (hunk 12 above)
Comment (amiller68) [reply src/eth.rs:361]: it reduces code bloat

File: src/eth.rs (hunk 11 above)
Comment (amiller68) [reply src/eth.rs:182]: done

File: src/eth.rs (hunk 12 above)
Comment (amiller68) [reply src/eth.rs]: its gone

File: src/eth.rs (hunk 5 above)
Comment (amiller68) [reply src/eth.rs]: deprected builder pattern, implemented new

File: src/eth.rs (hunk 13 above)
Comment (amiller68) [reply src/eth.rs:408]: This does not work -- testing with an empty contract I deployed. Working on it.

File: src/eth.rs (hunk 8 above)
Comment (amiller68) [reply src/eth.rs]: It works

File: src/eth.rs (hunk 13 above)
Comment (amiller68) [reply src/eth.rs:425]: Got deal Proposal posting working (assuming it is configured properly). Working on detokenizing onchaindeals with the getter x defined and then I can make this test more robust by pulling the created deal.

**Review by laudiacay — DISMISSED** [review]

**Comment by laudiacay** [comment]
i pushed changes from #68 over here

**Review by laudiacay — APPROVED** [review]
yop

**Pull request merged**